_posts/2019-07-11-JS-testing-101.md

```diff
@@ -0,0 +1,433 @@
+This is the easiest way to test this scenario, although, it is a little verbose. To reduce the lines of code, a simple scenario like this, can also be tested by transforming the observable to a promise and test it like we did in the promise scenario:
+
+````javascript
+it('should get the cars where the color is red (alternative way)', async done => {
+      when(carService.getCarsAsObservable()).thenReturn(of([aRedPorsche, aBlackCadillac, aBrownCar]));
+
+      try {
+        const actual = await componentUnderTest.getAllRedCarsAsObservable().toPromise();
+        expect(actual).toEqual([aRedPorsche]);
+        done();
+      } catch (e) {
+        fail(e);
+        done();
+      }
+});
+
+````
+
+I guess in this case it depends on what you find the most readable. The async await syntax makes the second version very readable, but comes down to personal preference.
+
+#### Scenario B
+
+In the next scenario we will be testing an observable call with side effects. 
+
+````javascript
+getAllRedCarsAsObservablewithStuffHappeningInBetween(): void {
+    this.loading = true;
+
+    this.getAllRedCarsAsObservable()
+      .pipe(
+        finalize(() => this.loading = false)
+      )
+      .subscribe(
+        (cars: Array<Car>) => this.cars = cars,
+        () => this.carRetrievalError = true
+      );
+  }
+````
+
+Same scenario as the promises, we set loading to true and subscribe to an observable, when it resolves, we set the value and when it completes, we set loading to false. If the observable throws an error, we set carRetrievalError to true.
+
+So our first test is the happy path, the observable resolves with an array of cars and everything goes as expected:
+
+````javascript
+it('should set all the red cars to the cars field and set loading field correctly', async done => {
+    const carsSubject = new Subject<Array<Car>>();
+    when(carService.getCarsAsObservable()).thenReturn(carsSubject);
+
+    componentUnderTest.getAllRedCarsAsObservablewithStuffHappeningInBetween();
+
+    expect(componentUnderTest.loading).toBeTruthy();
+
+    carsSubject.next([aBlackCadillac, aRedPorsche]);
+    carsSubject.complete();
+
+    expect(componentUnderTest.cars).toEqual([aRedPorsche]);
+    expect(componentUnderTest.loading).toBeFalsy();
+    done();
+});
+````
+
+This scenario is very easily tested with observables, because of Subjects. What is a subject? A subject can act as both an observable and an observer. This means we can subscribe to it and receive the emitted values, or we can tell it to emit a certain value.
+In this case, instead of telling the carService to return an Observable that immediately resolves with a value, we tell it to return a subject.asObservable(). Now we are in control of when and what value is emitted in this observable. This allows us to check our side effects easily at the right time.
```

Comment:
```suggestion
In this case, instead of telling the `carService` to return an observable that immediately resolves with a value, we tell it to return a `subject.asObservable()`. Now we are in control of when and what value is emitted in this observable. This allows us to check our side effects easily at the right time.
```